custom/shared/hooks/useSharedState.js

```diff
@@ -0,0 +1,112 @@
+import React, { useState, useEffect, useCallback, useRef } from 'react';
+import { useCallState } from '../contexts/CallProvider';
+
+// @params initialValues - initial values of the shared state.
+// @params broadcast - share the state with the other participants whenever the state changes.
+export const useSharedState = ({ initialValues = {}, broadcast = true }) => {
+  const { callObject } = useCallState();
+  const stateRef = useRef(null);
+
+  const [state, setState] = useState({
+    sharedState: initialValues,
+    setAt: undefined,
+  });
+
+  // handling the app-message event, to check if the state is being shared.
+  const handleAppMessage = useCallback(
+    event => {
+      // two types of events -
+      // 1. Request shared state (request-shared-state)
+      // 2. Set shared state (set-shared-state)
+      switch (event.data?.message?.type) {
+        // if we receive a request-shared-state message type, we check if the user has any previous state,
+        // if yes, we will send the shared-state to everyone in the call.
+        case 'request-shared-state':
+          if (!stateRef.current.setAt) return;
+          // random delay to avoid too many messages being sent at the same time from different clients.
+          const randomDelay = 200 + Math.ceil(1000 * Math.random());
+          setTimeout(() => {
```

Comment:
This timeout is probably not needed. We can immediately send the state.

Reply:
Was using the timeout because we are trying to get the state from every user in the call, so it might cause the local participant to receive state at the same time.

If we change to get state from only 1 participant then we can remove that timeout, I agree.